crypto_box.js

```diff
@@ -95,6 +103,159 @@ function crypto_box_seal_open (m, c, pk, sk) {
   return crypto_secretbox_open_easy(m, c.subarray(epk.length), n, k)
 }
 
+// int
+// crypto_box_curve25519xsalsa20poly1305_beforenm(unsigned char *k,
+//                                                const unsigned char *pk,
+//                                                const unsigned char *sk)
+// {
+//     static const unsigned char zero[16] = { 0 };
+//     unsigned char s[32];
+//
+//     if (crypto_scalarmult_curve25519(s, sk, pk) != 0) {
+//         return -1;
+//     }
+//     return crypto_core_hsalsa20(k, zero, s, NULL);
+// }
+function crypto_box_beforenm (k, pk, sk) {
+  const zero = new Uint8Array(16)
+  const s = new Uint8Array(32)
+
+  if (crypto_scalarmult(s, sk, pk) !== 0) {
+    return -1
+  }
+
+  return xsalsa20.core_hsalsa20(k, zero, s, null)
+}
+
+// int
+// crypto_box_detached_afternm(unsigned char *c, unsigned char *mac,
+//                             const unsigned char *m, unsigned long long mlen,
+//                             const unsigned char *n, const unsigned char *k)
+// {
+//     return crypto_secretbox_detached(c, mac, m, mlen, n, k);
+// }
+function crypto_box_detached_afternm (c, mac, m, n, k) {
+  return crypto_secretbox_detached(c, mac, m, n, k)
+}
```

Comment:
Is it best to copy the exact call stack that libsodium uses, or should I have just put this call in `crypto_box_detached`?

Reply:
this works well here. when in doubt we favour clarity over conciseness, as does this PR :)